Feature/dm 191 refactor load processed pileup

OberonDixon opened the pull request:
## Description

The core of this PR is a set of adjustments to load_processed for pileup vectors, pileup counts, and regions_to_list, with the goal of utilizing multiple cores to speed up processing and reduce memory overhead. Similar changes are possible conceptually for the single read loaders, but given that we may adjust single read loading to a new file type with new modkit versions and given that I have mostly had performance bottlenecks in data loading with pileups, I opted to keep that as-is. Most changes are directly related to this goal; some minor things were incidentally updated because I was overhauling that part of the code anyway.

Here is an overview of the changes by file:

### utils.py

1. Added process_chunks_from_regions_dict, which lets you break up a regions_dict (regions by chromosome) into a region chunks dict, with subregions of the specified size. This helps with parallelization later by providing a shareable function that chunks stuff up.

### load_processed.py

1. The whole file was reorganized into sections: Load wrappers (which let you call other loaders for many regions simultaneously), Pileup loaders, and Single read loaders. Comments were added to function descriptions indicating which are User-facing as opposed to internal helpers.
2. For the currently-implemented load wrapper function, regions_to_list (used for read depth histograms currently but applicable for a bunch of stuff later) was updated to either parallelize across the different load calls OR parallelize within the load calls, based on whether parallelize_within_region is True. This is useful if one is analyzing several large regions rather than many, many small regions. Logic was added to handle progress bars.
3. The process_region function, called by regions_to_list, was shifted down to follow the organization paradigm of top-level loader first, helpers after
4. pileup_counts_from_bedmethyl was updated to enable non-windowed loading, which is convenient for certain analysis tasks (NOT essential for parallelization)
5. pileup_counts_from_bedmethyl and pileup_vectors_from_bedmethyl were both adjusted to reference the same shared row parsing logic, now inside process_pileup_row. 
6. pileup_counts_from_bedmethyl and pileup_vectors_from_bedmethyl got progress bars, that can be turned off with quiet=True. 
7. pileup_counts_from_bedmethyl and pileup_vectors_from_bedmethyl both got their more specific logic moved into respective process_chunk functions, which are what get run by the parallel process executor for chunks. Experimentation showed that a chunk size of 1 million bp is fastest in pretty much every case: for small regions, that is just the whole region, whereas for whole-chromosome processing you get good resource utilization, relatively low memory footprint, and overall fast speed. 
8. The output values for pileup_counts_from_bedmethyl and pileup_vectors_from_bedmethyl are handled with multiprocessing shared_memory.SharedMemory objects, which let you share a range of memory (that I map to numpy arrays - other mappings such as to int proved challenging to maintain the mapping to shared memory for editing) between all of the cores. With a lock that is also shared, we avoid conflicting write attempts. There is some nuance in handling these objects properly; calling .close() on the instantiated reference in the child process function seems crucial and then .close() *followed by* .unlink() is absolutely crucial in the manager function to avoid catastrophic failure. 
9. In the single read loader section, some additional comments were added describing functionality where it was missing

### export.py

1. Updated the bigwig exported to reference the new load_processed.process_pileup_row function to use shared logic with load_processed pileup handlers. This cleans up quite a lot, it turns out.

### parse_bam.py

1. ruff demanded tiny format changes so I relented.

### plot_depth_histogram.py / plot_depth_profile.py / plot_enrichment.py / plot_enrichment_profile.py

1. Added passthrough parameter for keeping pbars quiet

### run_modkit.py

1. ruff demanded tiny format changes so I relented.

### cases.py

1. Added test coverage for different single_strand, regions_5to3prime, chunk_size, and cores values. These were all important to test before and after the refactor because I totally rewrote the core strand handling logic, and the cores parameter is now dealing with things that we implemented as opposed to modkit. 

### test data bed files

1. Added strand info per the actual bed spec, i.e. in the right column, so the tests actually test something.

### dimelo_test.py

1. Explained decision to override cores test value for extract specifically
2. Some formatting changes demanded by ruff

### generate_targets.py

1. New logic to properly handle test case management when trying to adjust some tests to increase coverage while wanting to keep other test results identical, including generation timestamps, for the purposes of easy traceability of time-of-validity. 
2. Added logic to remove the cores parameters passed to extract at generation time, per reason explained in dimelo_test.py i.e. because of the non-deterministic nature of modkit extract in terms of ordering (at least as of 0.2.4)

## Things to Focus On

The core logic does seem to work identically to the previous implementation, as shown by my test infrastructure upgrades on a separate branch without the refactor. Certainly any improvements to clarity of the code would be great, because it has become in some ways less readable/clear. We want to make sure comments are sufficient everywhere and that it all flows nicely. Also, is it right to show pbars by default? Maybe not. Open to other interface adjustments too.

Commit:
DM-229 Consolidating parsing logic for load_processed functions. Still need to apply to export functions, and need to check that performance is unimpacted.

Commit:
DM-231 Removed regions=None case for pileup_counts_from_bedmethyl because it will complicate parallelization and is not clearly useful.

Commit:
DM-229 Adjusted export.pileup_to_bigwig to use new load_processed.process_pileup_row and renamed/configured a few variables to support this.

Commit:
DM-192 Implemented parallelization core including shared memory and chunk generation, currently only for load_processed.pileup_vectors_from_bedmethyl. pileup_counts_from_bedmethyl is a simpler case. Passes pre-existing tests including when chunk_size<region size, but further infrastructure is required to assess whether regions_5to3prime still works right (logic gets a bit complicated). Speed with a single core and a small number of small regions is slightly slower due to overhead necessary to support parallelization. Need to test that we get a speedup with multi-core long regions or many regions.

Commit:
DM-232 Added strand information to regions .bed files. Re-generated read_vectors_from_hdf5 test targets to pass tests; other test targets are the same for now.

Commit:
DM-232 Added single_strand and regions_5to3prime to cases.py. Adjusted generate_targets.py to properly handle copying over new kwargs from cases.py when re-running some but not all targets: if all targets are re-run, then the old test_matrix.pkl is discarded. If only some are re-run, we need the old test_matrix.pkl, but still need to copy over all the kwargs for all cases that won't be re-run and still need to copy over old results, which will be overridden for re-run targets.

OberonDixon committed:
Merge pull request #16 from streetslab/feature/DM-191-refactor-load-processed-tests-baseline

Feature/dm 191 refactor load processed tests baseline

Commit:
DM-192 Fixed logic to properly handle single_strand case within process_pileup_row. Passing all tests.

Commit:
DM-192 Further refactor to simplify process_pileup_row and keep regions_5to3prime logic within the pileup_vectors_from_bedmethyl function, which is the only one that actually needs it. Runs faster, still passes all tests.

Commit:
DM-192 Added chunk_size for load_processes parallelization to cases.py, re-ran targets for pileup_counts_from_bedmethyl and pileup_vectors_from_bedmethyl.

OberonDixon committed:
Merge pull request #17 from streetslab/feature/DM-191-refactor-load-processed-tests-baseline

DM-192 Added chunk_size for load_processes parallelization to cases.p…

Commit:
DM-192,DM-236,DM-233,DM-235 Parallelized pileup_counts_from_bedmethyl. Added progress bars and quiet toggles for pileup loaders. Set default chunk size to 1 MB after whole-chromosome speed testing. Added option to parallelize within rather than between regions with regions_to_list.

Commit:
DM-192,DM-236 Properly close and delink shared memory when appropriate to avoid intermittent crashes. Adjust tabix fetch to avoid negative start values. Tweak wording on progress bars.

Commit:
DM-191 Re-arrange contents of load_processed for better clarity. Added some comments.

Commit:
DM-238 Add quiet and cores parameters and corresponding documentation to all pileup plotters. A few other small tweaks.

Commit:
DM-239 Added test case coverage for 1-4 cores. GitHub jobs for Ubuntu and Macos ostensibly have 4 cores. This should not change any outputs because for extract the test is hardcoded to use 1 core, and that is the only one with any stochasticity in parallelization.

Commit:
DM-239 Added cases.py support for cores=None. Tweaked dimelo_test.py to avoid ever sending the cores arg twice, and added a comment to explain the reason why extract is tested unparallelized, unlike other functions.

Commit:
DM-239 Adjusted generate_targets module to properly handle cores argument for extract. Updated test_matrix.pickle by re-running generate_targets. No changes to pileup or extract files, but should pass all tests.

Commit:
Ruff format fixes

OberonDixon committed:
Merge pull request #18 from streetslab/feature/DM-191-refactor-load-processed-tests-baseline

New test_matrix and cases to cover 1, 2, 3, 4, and None for cores

OberonDixon committed:
Merge branch 'main' into feature/DM-191-refactor-load-processed

Commit:
Fixed small mistakes from merge

OberonDixon committed:
Merge branch 'main' into feature/DM-191-refactor-load-processed

thekugelmeister commented on dimelo/plot_enrichment.py:
```diff
@@ -31,6 +33,8 @@ def plot_enrichment(
         window_size: (currently disabled) window around center of region, +-window_size//2
         single_strand: True means we only grab counts from reads from the same strand as
             the region of interest, False means we always grab both strands within the regions
+        quiet: disables progress bars
```
Neither of these arguments are passed down to `get_enrichments`

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -11,38 +15,35 @@
 
 def regions_to_list(
     function_handle,
     regions,
```
Can you add a type hint to this? Does it expect all of the usual suspects for regions, or just a subset of them like a bam file?

OberonDixon replied:
Clarified in comment. Any format is fine that is supported elsewhere.

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -11,38 +15,35 @@
     window_size: int | None = None,
+    quiet: bool = True,
     cores: int | None = None,
+    parallelize_within_regions: bool = False,
```
Why is this option here? Do you have any use cases in mind where this would be useful?

OberonDixon replied:
Clarified use case in docstring. It is for small numbers of large regions, e.g. comparing chromosomes.

thekugelmeister replied:
I loosely understand the goal of the use case, but I find the wording difficult to parse and I'm not sure I fully follow the logic.

If I'm understanding correctly, the goal is to specify whether to split up large regions into smaller chunks. If you have small regions, you can just be embarrassingly parallel across regions. If you have larger ones, you want to split those regions up. Is that high level accurate?

If this understanding is true, then is the current configuration that `False` --> don't split regions and `True` means to split regions?

If all of this is true, I suggest maybe a rename of this argument to `split_large_regions` or something like that? What do you think?

I only care because this is user facing and I find the current name/docs a little arcane. If I'm understanding correctly, this is an important setting to include, so let's make it clear to the user why they should care.

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -77,27 +90,54 @@ def regions_to_list(
     return results
 
 
+def process_region(region_string, function_handle, **kwargs):
```
I'm probably misunderstanding, but why is this a separate method? Why can't it be part of the `partial` above? If it needs to exist, I think it needs a better name; this one is very generic. Can you describe more fully what it is intended to do?

OberonDixon replied:
I think the executor.map call needs a positional input rather than a keyword (e.g. regions), but then the ones that take keywords don't seem to be happy to return things in the same order as they were sent (necessary for the outputs list to map to the input regions), so I am using this wrapper. Renamed it for clarity to apply_loader_function_to_region. Happy to use a different multiprocessing approach that can directly handle kwargs if you know of one.

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -77,27 +90,54 @@ def regions_to_list(
     bedmethyl_file: str | Path,
     motif: str,
-    regions: str | Path | list[str | Path] | None = None,
+    regions: str | Path | list[str | Path],
```
Is the idea that nobody would ever ask for a count over the entire genome? Why did this option go away?

thekugelmeister replied:
Or is this a byproduct of the parallelization that I do not understand?

OberonDixon replied:
If we want to be able to parallelize across regions, that can't be done genome-wide without extra work to delineate what will be processed by which worker. I also think that the use case for whole-genome pileup counts is very narrow, and moreover, you just need to past a bed file of chromosomes if that's what you want. I think allowing None here honestly is most likely to result in confusion rather than support a coherent use case.

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -77,27 +90,54 @@ def regions_to_list(
-    cores: int | None = None,  # currently unused
+    quiet: bool = False,
+    cores: int | None = None,
+    chunk_size: int = 1_000_000,
```
Make a `DEFAULT_CHUNK_SIZE` global variable at the top of the file and reference here (and in other `chunk_size` defaults)

thekugelmeister commented on dimelo/utils.py:
```diff
@@ -103,6 +103,28 @@ def adjust_threshold(
     return thresh
 
 
+def process_chunks_from_regions_dict(
```
At minimum, annotate with return type or a description of what is returned

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -118,101 +158,72 @@ def pileup_counts_from_bedmethyl(
-    return (modified_base_count, valid_base_count)
+    cores_to_run = utils.cores_to_run(cores)
 
+    shm_valid = shared_memory.SharedMemory(
```
Can the processing code in here and `pileup_vectors_from_bedmethyl` be combined into a single helper method? It feels as if there's a ton of shared functionality, and it clutters up what the top-level methods are doing.

OberonDixon replied:
I think I would prefer not to break these out, because where the code is shared it is already very simple and as close to self-documenting as I think we could achieve, and where the code is subtly different it would not really be feasible to parameterize so as to support any future use cases, unless we just passed down the size of the SharedMemory buffer directly and allowed the user to request an arbitrary number of shared memory buffers of that size, which I think is pointless.

OberonDixon replied:
I added some comments to better clarify what is happening with the variable setup

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -254,86 +269,100 @@ def pileup_vectors_from_bedmethyl(
+
 def vector_from_fake(window_size: int, *args, **kwargs) -> np.ndarray:
     """
+    Tes helper function.
```
Typo

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -347,6 +376,234 @@ def vector_from_fake(window_size: int, *args, **kwargs) -> np.ndarray:
+    existing_valid.close()
+
+
+def pileup_counts_process_chunk(
```
Similar question to above, although maybe I'm picking up that this is because the summation is being done in here instead of at the end? Which probably makes sense for memory efficiency.

If this is correct, maybe these two methods can stick around for now, but the way in which they are called can be consolidated?

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -376,6 +636,8 @@ def read_vectors_from_hdf5(
 
     After this processing, we calculate modification fractions, sort, and return.
 
+    TODO: Implement progress bars and parallelization as with pileup loaders
```
Wait, isn't this implemented? Oh or is this why I was confused about why things took a while? I'd have to go check. Ask me about this when you read this.

OberonDixon replied:
Not for single reads it isn't.

thekugelmeister commented on dimelo/load_processed.py:
```diff
@@ -604,10 +867,13 @@ def readwise_binary_modification_arrays(
     subset_parameters: dict | None = None,
 ) -> tuple[list[np.ndarray], np.ndarray[int], np.ndarray[str], dict | None]:
     """
+    Primarily designed as a helper function for single-read plotting, but can be used by a user.
```
...yeah, I guess? Is the clarification helpful?

thekugelmeister commented on dimelo/export.py:
```diff
@@ -134,54 +134,31 @@ def pileup_to_bigwig(
-                elif len(pileup_basemod.split(",")) == 3:
-                    pileup_modname, pileup_motif, pileup_mod_coord = (
-                        pileup_basemod.split(",")
+                keep_basemod, genomic_coord, modified_in_row, valid_in_row = (
```
Very much appreciate the simplification!

thekugelmeister commented on dimelo/plot_enrichment_profile.py:
```diff
@@ -41,6 +43,8 @@ def plot_enrichment_profile(
             the region of interest, False means we always grab both strands within the regions
         regions_5to3prime: True means negative strand regions get flipped, False means no flipping
         smooth_window: size of the moving window to use for smoothing. If set to None, no smoothing is performed
+        quiet: disables progress bars
```
Neither of these arguments are passed down to `get_enrichment_profiles`

thekugelmeister reviewed (changes requested):
Anecdotal testing comments:
* A figure that took me 30 minutes to produce this morning took me 15 minutes this afternoon, on a not very optimized setup. My computer might be screaming in pain but I can adjust that. I'm theoretically quite content on the whole with that result, although I'm curious if you've seen better speedups on savio.
* Other plots have had amazing speedups since. Still not sure what that was about.

General comments from review:
* I feel like we should add clarification of what the options to `cores` can be, for the user-facing methods. I understand that they all are the same (`None` means "all available", etc), and it's slightly redundant to say it everywhere, but I think it would be useful for the end users.
* Progress bars
  * Almost every time I talk to a new package user, they say, "it's nice it has progress bars" 🤣 That being said, we don't need them to stick around after the fact. I might actually start arguing that _all_ of our progress bars should be set to `leave=False` or `=None`. 
  * I think the message on the bar should be changed to something more generic like "loading data". I think it will suffice and look cleaner.
* The fact that it prints `modkit found with expected version 0.2.4` for every core is mildly annoying. Now that I think about it, is it necessary to print that at all? Can we just have it print if none or the wrong version is found?

Commit:
Added check for chrom in tabixfile.

OberonDixon commented:
## Performance test results

These are the speedups I see on Savio. The second graph is showing the results for fewer cores also, so should tell also what to expect for e.g. an 8-core macbook.

![image](https://github.com/user-attachments/assets/b028224a-f847-41d4-8f24-8610ac1f99dd)
![image](https://github.com/user-attachments/assets/0371629a-8eca-457b-aff2-168b3da25d22)

## High level feedback response

I got rid of the successfully-found-modkit printout, leaving the error outputs. Adjusted the progress bars to disappear when done for load_processed (not for parse_bam though we can do that in the future). Changed description to just "Loading data."

## Granular feedback response

I made changes to fix most of the requests. I decided not to consolidate the SharedMemory management because there are subtle differences in how the variables are created between the counts and vectors method (added comments describing what they are doing), and the dereferencing code is so straightforward already. 

## Other changes

I added this to both the vectors and counts loader. I had that in previous versions of these pileup loaders but removed for simplicity, only to realize/remember that we get errors if we don't make sure the chromosomes is present in the tabix index before fetching.
```
        # tabix throws and error if the contig is not present
        # by the current design, this should be silent
        if chromosome in source_tabix.contigs:
            for row in source_tabix.fetch(
                chromosome, max(subregion_start, 0), subregion_end
            ):
```

Commit:
DM-191 Adjustments in response to PR comments from thekugelmeister

Commit:
Merge branch 'feature/DM-191-refactor-load-processed' of https://github.com/streetslab/dimelo_v2 into feature/DM-191-refactor-load-processed

thekugelmeister reviewed (changes requested):
Looks almost good! I requested some additional clarification in one place for a user-facing argument but everything else i think looks fine.

Commit:
Adjust naming for regions_to_list region-splitting parallelization parameter.

thekugelmeister committed:
attempting to fix ruff versioning

thekugelmeister committed:
ruff version fix attempt 2
